### ISSUE TITLE
With overflow:clip and line-clamp CSS properties, users of AT have a very different content experience than sighted users have

### ISSUE THREAD

**Issue by thejohnjansen** [w3c/css-aam#29]
When authors create clipped or clamped content, sighted users have limited access to the content of those styled elements, but the a11y tree and the DOM see the full content; therefore, assistive technologies will read out all of the text. 

[I have created this example](https://codepen.io/editor/john-jansen/pen/019ff1c4-e7a4-724f-b136-abb992c6a437) to demonstrate the issue. The code has three identical paragraphs of text: one is clipped, one is clamped, and one is hidden. A sighted user only sees a portion of the paragraph; however, a user of a screen reader will hear all of the text read to them in the clip and clamp cases - even though some amount of text is not visible on the page.

It is [unclear](https://github.com/w3c/csswg-drafts/issues/12859) to the CSS working group how these styles should be treated by the AT. I took the action item at the CSSWG face-to-face to log this issue in order to get clarity.

I have been unable to find tests that explicitly demonstrate this issue.

**Comment by jnurthen** [comment]
This was discussed in the ARIA WG on Aug 20 2026

<Zakim> agendum 5 -- -> With overflow:clip and line-clamp CSS properties, users of AT have a very different content experience than sighted users have
<Zakim> ... https://github.com/https://github.com/w3c/css-aam/issues/29 -- taken up [from agendabot]
<front-endian-jane> s/@jn PR 2869 needs another reviewer/ /
<front-endian-jane> jamesn: We briefly discussed this in triage last week
<giacomo-petri> q+
<HaTheo> q+
<jamesn> q+
<front-endian-jane> JohnJansen: I added 3 examples to the github issue. We want confirmation and tests written that the way we want this to behave is that truncation isn't read by assistive technologies, the full item is read even when a truncation is rendered to the screen.
<jamesn> Ack giacomo-petri
<aardrian> Q+
<front-endian-jane> giacomo-petri: There is often a "more" button to see all the text, but for screen reader users that is useless because everything is already readable to them. Maybe there should be a CSS property to also hide the text to assistive technologies?
<front-endian-jane> JohnJansen: Interesting proposal, I can file an issue for it, but that is orthogonal to this particular issue.
<Matt_King> q+
<jamesn> Ack HaTheo
<front-endian-jane> JohnJansen: The current behavior of everything being read by ATs is how it works in every browser and has many reasons to work that way. I need an expert to confirm that it is supposed to work that way and we should write a test to enforce that.
<jamesn> Ack jamesn
<front-endian-jane> HaTheo: Yes, we very much want it to work this way. Some people use screen readers and zoom the text in at the same time. It's annoying to have to click a button to read the rest as a screen reader user. there are some corner cases that have to be thought of but I don't recommend changing this.
<front-endian-jane> jamesn: I don't think we could change because it's been this way for so long. Yes this is a good expected behavior. I also think it would be worth adding a way to hide truncated text as giacomo-petri recommended. You also get into weird situations when there are hover to show more interactions. We should think about how to make all that better.
<jamesn> Ack aardrian
<jamesn> q+ to loop back on the scrolling area thing
<front-endian-jane> aardrian: Largely I agree with what jamesn and HaTheo said, but issue mostly talks about screen reader experiences and concerns, but this impacts many other AT users. The issue should also note that some screen reader users are sighted. Should there be a separate conversation about making clipping keyboard accessible the same way scrollable regions are?
<jamesn> Ack Matt_King
<Daniel> s|issue filled for this|-> https://github.com/https://github.com/w3c/aria-practices/issues/3414 issue filled for this|
<front-endian-jane> Matt_King: I get that it has worked this way for a long time, but to me it feels like it causes a lot of problems for real screen reader users like me. I can't tell what parts are visible and what parts aren't. I can't tell that clipped content is being read to me in the examples given.
<front-endian-jane> JohnJansen: I can fix that in the examples.
<front-endian-jane> Matt_King: The clipped content can be anything, even interactive elements, right?
<front-endian-jane> JohnJansen: Correct.
<jamesn> Q-
<HaTheo> yes the clip might even be the name of a button.
<front-endian-jane> Matt_King: This is where I've run into problems, when the volume of content is large and the reasons it is clipped is it's in a collection of some kind. It makes it hard to skip to the content you want to get to that isn't clipped since you have to wade through all the clipped content. I canm
<jamesn> Q?
<front-endian-jane> s/I canm/I can't tell when it is or isn't clipped so I don't always know when this is causing problems/
<HaTheo> +1 to giving authors a way of clipping that might be more friendly and better.
<front-endian-jane> q+
<jamesn> q+
<front-endian-jane> JohnJansen: I can take all this back to the group, and we need to think about having some way to hide the more button with a CSS property, We also may need a way to make it possible to hide clipped content to ATs for some cases.
<jamesn> Ack front-endian-jane
<Matt_King> q+
<jamesn> Q-
<jamesn> Ack Matt_King
<aardrian> pretend "keyboard user" includes voice users as well
<jamesn> Switch user too
<front-endian-jane> front-endian-jane: There could be a lot of issues for people who are sighted screen reader users if we give an option to hide things like a "more" button when content is clipped but read because they would lose AT access to the more button, but still see it and may still want to click it.
<giacomo-petri> +1 to Matt_King
<giacomo-petri> q+
<front-endian-jane> Matt_King: We definitely want a way to clip content for AT users too since often content is being clipped to reduce how busy the page is, reduce cognitive overload, etc.
<jamesn> Ack gia
<front-endian-jane> jamesn: There is definitely a problem to be solved that isn't solved today.
<front-endian-jane> Matt_King: We want to think of clipping as not just a visual thing.
<jamesn> agenda?
<front-endian-jane> giacomo-petri: Having an additional property may be essential for authors to make good show/hide UI's and to properly control their UX in the way they intend. A lot to think about on this.
<front-endian-jane> zakin, next item
<front-endian-jane> zakim, next item

**Comment by thejohnjansen** [comment]
Per the above conversation, I have update my examples found here: https://codepen.io/editor/john-jansen/pen/019ff1c4-e7a4-724f-b136-abb992c6a437

**Comment by jnurthen** [comment]
Summarized consensus from above:

* Current behavior is correct and should be preserved. Clipped/truncated content being read in full by ATs is how every browser works today, is the expected behavior, and shouldn't change — it's been this way too long, and users who combine a screen reader with zoom rely on it. Tests should be written to enforce it.
* But there's a real unsolved problem. Matt_King noted he can't tell which content is visible vs. clipped, which makes large clipped collections hard to navigate. Clipped content can include interactive elements — even a button's name. Broad agreement that clipping should be treated as more than a visual concern, since it's often used to reduce cognitive load, and AT users deserve that same benefit.
* A CSS mechanism to hide clipped content (and the "more" button) from ATs is worth pursuing but as a separate issue, orthogonal to the current one. ARIA-practices issue for this filed as [w3c/aria-practices#3414](https://github.com/w3c/aria-practices/issues/3414).

**Comment by jnurthen** [comment]
@thejohnjansen please re-add agenda to this once discussion has taken place in CSS WG
